flaky test: TestForceExec — 1205 when the force-kill lands inside the ~100ms budget left by LockWaitTimeout=1 (no post-kill retry)

## What happened

`TestForceExec` (`pkg/dbconn/dbconn_test.go`) failed on PR #960 in the MySQL 8.0.28 (Aurora version) workflow:

```
INFO found locking transaction pid=1077 lockType=SHARED_READ ... objectName=requires_mdl runningTime=2.09s
INFO found locking transactions count=1 pids=[1077]
WARN killing locking transaction pid=1077
dbconn_test.go:291: Received unexpected error:
    Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
[mysql] packets.go:58 unexpected EOF
--- FAIL: TestForceExec (2.14s)
```

Failing run: https://github.com/block/spirit/actions/runs/27389248718/job/80942995467

The kill machinery worked — it found and killed the blocking transaction (the `unexpected EOF` is the killed connection) — but `ForceExec`'s ALTER hit its lock-wait timeout anyway, so the final `require.NoError` failed.

## Why

The test sets `config.LockWaitTimeout = 1` ("as short as possible"). `ForceExec` schedules `KillLockingTransactions` via `time.AfterFunc(forceKillGracePeriod(...))`, and `forceKillGracePeriod(1)` = max(1 × 0.9, 0.9) = **0.9s** — while the server expires the ALTER's MDL wait at **1.0s**.

That leaves a ~100ms budget for: the timer goroutine to be scheduled, the performance_schema locking-transaction scan to run, the KILL to execute, and the server to release the MDL and grant the waiting ALTER — all on a CI host that is concurrently running the heavy copy tests (this failure landed in the same loaded window as the #946 recurrence and #963). When any of those steps takes >100ms, the ALTER gets error 1205 just before/as the kill lands, and `ForceExec` returns the error — it makes a single attempt with **no retry after the kill fires**.

At the production default `LockWaitTimeout` (e.g. 30s → kill at 27s, ~3s budget) the race is academic; it's the test's minimum setting that exposes it.

## Why it's not the PR under test

PR #960 touches `pkg/change`/`pkg/datasync`/`pkg/move`/`pkg/table` — no `pkg/dbconn` code.

## Suggested fixes (either suffices)

- **Test-level:** use `LockWaitTimeout = 2` (kill at 1.8s, ~200ms budget) or 3, trading a slightly slower test for a budget that survives loaded runners.
- **Code-level (more robust for production too):** if the protected statement fails with error 1205 *and* the kill timer already fired, retry the statement once — the blocker is now dead, so the retry acquires the MDL immediately. This also covers the real-world case where the kill lands a hair too late at any timeout setting.

--- a/pkg/dbconn/dbconn.go
+++ b/pkg/dbconn/dbconn.go
@@ -11,6 +11,7 @@ import (
 	"log/slog"
 	"math/rand/v2"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/block/spirit/pkg/dbconn/sqlescape"
@@ -317,15 +318,18 @@ func ForceExec(ctx context.Context, db *sql.DB, tables []*table.TableInfo, dbCon
 	// since the minimum LockWaitTimeout=1 second
 	duration := forceKillGracePeriod(dbConfig.LockWaitTimeout)
 	var wg sync.WaitGroup
+	var killTimerFired atomic.Bool
 	wg.Add(1)
 	timer := time.AfterFunc(duration, func() {
 		defer wg.Done()
+		killTimerFired.Store(true)
 		err := KillLockingTransactions(ctx, db, tables, dbConfig, logger, []int{connId})
 		if err != nil {
 			return // just return, we can't do much more here
 		}
 	})
-	_, err = trx.ExecContext(ctx, sqlescape.MustEscapeSQL(stmt, args...))
+	escapedStmt := sqlescape.MustEscapeSQL(stmt, args...)
+	_, err = trx.ExecContext(ctx, escapedStmt)
 	if timer.Stop() {
 		// Timer was stopped before it fired, so the goroutine never started.
 		// We need to manually decrement the WaitGroup.
@@ -335,7 +339,19 @@ func ForceExec(ctx context.Context, db *sql.DB, tables []*table.TableInfo, dbCon
 	// This prevents a race where the goroutine kills connections that
 	// are now being used for subsequent operations.
 	wg.Wait()
+	if shouldRetryForceExecAfterKill(err, killTimerFired.Load()) {
+		logger.Warn("retrying statement after lock wait timeout because force-kill timer fired", "error", err)
+		_, err = trx.ExecContext(ctx, escapedStmt)
+	}
 	return err
+}
+
+func shouldRetryForceExecAfterKill(err error, killTimerFired bool) bool {
+	if !killTimerFired {
+		return false
+	}
+	val, ok := errors.AsType[*mysql.MySQLError](err)
+	return ok && val.Number == errLockWaitTimeout
 }
 
 // Exec is like db.Exec but only returns an error.

--- a/pkg/dbconn/dbconn_test.go
+++ b/pkg/dbconn/dbconn_test.go
@@ -280,6 +280,16 @@ func TestRetryableTrxRetriesKilledConnection(t *testing.T) {
 	testRetryableTrxSurvivesKill(t, "retry_kill_conn", "KILL %d")
 }
 
+func TestShouldRetryForceExecAfterKill(t *testing.T) {
+	require.False(t, shouldRetryForceExecAfterKill(nil, true))
+	require.False(t, shouldRetryForceExecAfterKill(&mysql.MySQLError{Number: errLockWaitTimeout}, false))
+	require.False(t, shouldRetryForceExecAfterKill(&mysql.MySQLError{Number: errDeadlock}, true))
+	require.False(t, shouldRetryForceExecAfterKill(errors.New("not a mysql error"), true))
+	require.True(t, shouldRetryForceExecAfterKill(&mysql.MySQLError{Number: errLockWaitTimeout}, true))
+	require.True(t, shouldRetryForceExecAfterKill(fmt.Errorf("force exec failed: %w",
+		&mysql.MySQLError{Number: errLockWaitTimeout}), true))
+}
+
 func TestForceExec(t *testing.T) {
 	config := NewDBConfig()
 	config.LockWaitTimeout = 1 // as short as possible.

